In [ ]:
import sys
from pathlib import Path

PROJECT_ROOT = Path().resolve().parent
sys.path.insert(0, str(PROJECT_ROOT))

from src.data.load_data import load_bank_data
from src.data.preprocess import split_data, BankFeatureEngineer, build_non_scaled_pipeline
from src.models.random_forest import (
    train_random_forest,
    find_best_threshold,
    evaluate_rf,
    save_random_forest,
)

In [2]:
#Training Block

# Load dataset
df = load_bank_data("../data/raw/bank-full.csv")

# Split
X_train, X_test, y_train, y_test = split_data(df)

# Detect engineered columns
fe = BankFeatureEngineer()
X_tmp = fe.fit_transform(X_train)

num_cols = X_tmp.select_dtypes(include=["int64", "float64"]).columns.tolist()
cat_cols = X_tmp.select_dtypes(include=["object", "category"]).columns.tolist()

# Build preprocessing pipeline
preprocess_pipeline = build_non_scaled_pipeline(num_cols, cat_cols)

# Train RF
best_rf, best_params, best_cv_f1 = train_random_forest(
    preprocess_pipeline,
    X_train,
    y_train
)

print("Best params:", best_params)
print("Best CV F1:", best_cv_f1)

Fitting 5 folds for each of 48 candidates, totalling 240 fits
Best params: {'classifier__class_weight': 'balanced', 'classifier__max_depth': 25, 'classifier__max_features': 0.5, 'classifier__min_samples_leaf': 10, 'classifier__min_samples_split': 2, 'classifier__n_estimators': 400}
Best CV F1: 0.4718180880292125


In [3]:
#Evalation Block and model saving

# Find optimal threshold
best_threshold, best_f1 = find_best_threshold(
    best_rf,
    X_test,
    y_test
)

print("Best threshold:", best_threshold)
print("Best F1:", best_f1)

# Final evaluation
metrics = evaluate_rf(
    best_rf,
    X_test,
    y_test,
    best_threshold
)

print("\n--- FINAL RANDOM FOREST ---")
for k, v in metrics.items():
    print(f"{k}: {v}")

# Save model
save_random_forest(best_rf, best_threshold)

Best threshold: 0.5247474747474747
Best F1: 0.4862710363153233

--- FINAL RANDOM FOREST ---
Accuracy: 0.8717239854030742
AUC: 0.7986759199964962
Precision: 0.4575
Recall: 0.5189035916824196
F1: 0.4862710363153233
MCC: 0.4143842505397125


### Random Forest Observations

- The dataset is highly imbalanced (~88% 'no', ~12% 'yes'), so F1-score is used as the primary evaluation metric.

- Random Forest is an ensemble learning method that combines multiple decision trees to reduce overfitting and improve generalization.

- Class imbalance was handled using class_weight="balanced".

- Hyperparameter tuning using GridSearchCV selected optimal values for max_depth, min_samples_leaf, and feature sampling.

- Compared to a single Decision Tree, Random Forest significantly improves stability and reduces variance.

- AUC score (~0.80) indicates strong ability to distinguish between positive and negative classes.

- The default threshold (0.5) was not optimal; threshold tuning found an optimal value ≈ 0.52.

- Final F1 score ≈ 0.486 shows the best performance so far among all models.

- Precision (~0.46) and Recall (~0.52) indicate a good balance:
  → Higher recall helps capture more positive cases
  → Precision remains reasonably controlled

- MCC (~0.41) confirms strong and balanced predictive performance.

Conclusion:
Random Forest outperforms simpler models by capturing non-linear relationships and reducing overfitting, making it one of the strongest models in this analysis.